In [303]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [304]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [305]:
# import excel file November2012.xlsx
tn_2012=pd.read_excel('../data/2012_election/November2012.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','BName1','Tally1','BName2','Tally2',
                               'BName3','Tally3','BName4','Tally4','BName5','Tally5','BName6','Tally6','BName7','Tally7',
                               'BName8','Tally8','BName9','Tally9','BName10','Tally10'])
tn_2012

COUNTY       PRECINCT                                   OFFICENAME  \
0      ANDERSON  Andersonville                      United States President   
1      ANDERSON  Andersonville                         United States Senate   
2      ANDERSON  Andersonville                         United States Senate   
3      ANDERSON  Andersonville                         United States Senate   
4      ANDERSON  Andersonville       U.S. House of Representatives District   
...         ...            ...                                          ...   
14087    WILSON    Provisional                         United States Senate   
14088    WILSON    Provisional                         United States Senate   
14089    WILSON    Provisional       U.S. House of Representatives District   
14090    WILSON    Provisional  Tennessee House of Representatives District   
14091    WILSON    Provisional  Tennessee House of Representatives District   

              ELECTDATE                       BName1  Tally1  \
0      November 6, 2012          . Mitt Romney - (R)    1110   
1      November 6, 2012           . Bob Corker - (R)    1067   
2      November 6, 2012      . Write-In - Larry Crim       0   
3      November 6, 2012  . Write-In - Mary D. Wright       0   
4      November 6, 2012    . Chuck Fleischmann - (R)     996   
...                 ...                          ...     ...   
14087  November 6, 2012      . Write-In - Larry Crim       0   
14088  November 6, 2012  . Write-In - Mary D. Wright       0   
14089  November 6, 2012          . Diane Black - (R)      36   
14090  November 6, 2012         . Mark A. Pody - (R)      16   
14091  November 6, 2012           . Susan Lynn - (R)      30   

                             BName2  Tally2                       BName3  \
0              . Barack Obama - (D)     267         . Virgil Goode - (C)   
1           . Mark E. Clayton - (D)     216         . Kermit Steck - (C)   
2      . Write-In - Gary Gene Davis       0  . Write-In - Allen Hoenicke   
3                               NaN       0                          NaN   
4          . Mary M. Headrick - (D)     282     . Matthew Deniston - (I)   
...                             ...     ...                          ...   
14087  . Write-In - Gary Gene Davis       0  . Write-In - Allen Hoenicke   
14088                           NaN       0                          NaN   
14089             . Pat Riley - (G)       9        . Scott Beasley - (I)   
14090                           NaN       0                          NaN   
14091                           NaN       0                          NaN   

       Tally3                           BName4  Tally4  \
0           3               . Jill Stein - (G)       0   
1          17          . Martin Pleasant - (G)      11   
2           0   . Write-In - De Wayne A. Jones       0   
3           0                              NaN       0   
4          48        . Write-In - Richard Hall       0   
...       ...                              ...     ...   
14087       0   . Write-In - De Wayne A. Jones       0   
14088       0                              NaN       0   
14089       9  . Write-In - Mary Amanda Keifer       0   
14090       0                              NaN       0   
14091       0                              NaN       0   

                                     BName5  Tally5  \
0          . Ross C. "Rocky" Anderson - (I)       0   
1                  . Shaun E. Crowell - (I)      18   
2                 . Write-In - Jacob Maurer       0   
3                                       NaN       0   
4      . Write-In - Lady J Jean Howard-Hill       0   
...                                     ...     ...   
14087             . Write-In - Jacob Maurer       0   
14088                                   NaN       0   
14089          . Write-In - Rachel Robinson       0   
14090                                   NaN       0   
14091                                   NaN       0   

                    

In [306]:
tn_2012.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14092 entries, 0 to 14091
Data columns (total 24 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   COUNTY      14092 non-null  object
 1   PRECINCT    14092 non-null  object
 2   OFFICENAME  14092 non-null  object
 3   ELECTDATE   14092 non-null  object
 4   BName1      14092 non-null  object
 5   Tally1      14092 non-null  int64 
 6   BName2      10660 non-null  object
 7   Tally2      14092 non-null  int64 
 8   BName3      8449 non-null   object
 9   Tally3      14092 non-null  int64 
 10  BName4      8207 non-null   object
 11  Tally4      14092 non-null  int64 
 12  BName5      7762 non-null   object
 13  Tally5      14092 non-null  int64 
 14  BName6      7059 non-null   object
 15  Tally6      14092 non-null  int64 
 16  BName7      6489 non-null   object
 17  Tally7      14092 non-null  int64 
 18  BName8      4210 non-null   object
 19  Tally8      14092 non-null  int64 
 20  BName9

In [307]:
# merge entries in all RNAME columns into new column "name"
tn_2012_namemelt=tn_2012.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['BName1','BName2','BName3','BName4','BName5','BName6','BName7','BName8','BName9','BName10'],value_name="candidate")
# drop column "variable" created by .melt
tn_2012_namemelt=tn_2012_namemelt.drop(['variable'],axis=1)                          
tn_2012_namemelt.head()

COUNTY       PRECINCT                              OFFICENAME  \
0  ANDERSON  Andersonville                 United States President   
1  ANDERSON  Andersonville                    United States Senate   
2  ANDERSON  Andersonville                    United States Senate   
3  ANDERSON  Andersonville                    United States Senate   
4  ANDERSON  Andersonville  U.S. House of Representatives District   

          ELECTDATE                    candidate  
0  November 6, 2012          . Mitt Romney - (R)  
1  November 6, 2012           . Bob Corker - (R)  
2  November 6, 2012      . Write-In - Larry Crim  
3  November 6, 2012  . Write-In - Mary D. Wright  
4  November 6, 2012    . Chuck Fleischmann - (R)

In [308]:
# merge entries in all TALLY columns into new column "vote_count"
tn_2012_pvtallymelt=tn_2012.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['Tally1','Tally2','Tally3','Tally4','Tally5','Tally6','Tally7','Tally8','Tally9','Tally10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2012_pvtallymelt=tn_2012_pvtallymelt.drop(['variable'],axis=1)
tn_2012_pvtallymelt.head()

COUNTY       PRECINCT                              OFFICENAME  \
0  ANDERSON  Andersonville                 United States President   
1  ANDERSON  Andersonville                    United States Senate   
2  ANDERSON  Andersonville                    United States Senate   
3  ANDERSON  Andersonville                    United States Senate   
4  ANDERSON  Andersonville  U.S. House of Representatives District   

          ELECTDATE  vote_count  
0  November 6, 2012        1110  
1  November 6, 2012        1067  
2  November 6, 2012           0  
3  November 6, 2012           0  
4  November 6, 2012         996

In [309]:
# add "vote_count" to tn_2016_namemelt
tally=tn_2012_pvtallymelt['vote_count']
tn_2012_namemelt=tn_2012_namemelt.join(tally)
tn_2012_namemelt.head(3)

COUNTY       PRECINCT               OFFICENAME         ELECTDATE  \
0  ANDERSON  Andersonville  United States President  November 6, 2012   
1  ANDERSON  Andersonville     United States Senate  November 6, 2012   
2  ANDERSON  Andersonville     United States Senate  November 6, 2012   

                 candidate  vote_count  
0      . Mitt Romney - (R)        1110  
1       . Bob Corker - (R)        1067  
2  . Write-In - Larry Crim           0

In [310]:
tn_2012_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140920 entries, 0 to 140919
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   COUNTY      140920 non-null  object
 1   PRECINCT    140920 non-null  object
 2   OFFICENAME  140920 non-null  object
 3   ELECTDATE   140920 non-null  object
 4   candidate   75348 non-null   object
 5   vote_count  140920 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 6.5+ MB


In [311]:
# lower case column names
tn_2012_namemelt.columns=tn_2012_namemelt.columns.str.lower()
tn_2012_namemelt.head(2)

county       precinct               officename         electdate  \
0  ANDERSON  Andersonville  United States President  November 6, 2012   
1  ANDERSON  Andersonville     United States Senate  November 6, 2012   

             candidate  vote_count  
0  . Mitt Romney - (R)        1110  
1   . Bob Corker - (R)        1067

In [312]:
tn_2012_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140920 entries, 0 to 140919
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   county      140920 non-null  object
 1   precinct    140920 non-null  object
 2   officename  140920 non-null  object
 3   electdate   140920 non-null  object
 4   candidate   75348 non-null   object
 5   vote_count  140920 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 6.5+ MB


#### 75348 non-null names  out of 140920 enties. Confirm all are 0 votes and drop rows with no candidate listed

In [313]:
# Create dataframe "blank_candidate_2012" of null valuea in "candidate" of tn_2012_namemelt
blank_candidate_2012=tn_2012_namemelt[tn_2012_namemelt['candidate'].isna()]
blank_candidate_2012

county       precinct                                   officename  \
14095   ANDERSON  Andersonville                         United States Senate   
14101   ANDERSON     Briceville                         United States Senate   
14107   ANDERSON       Bull Run                         United States Senate   
14113   ANDERSON        Clinton                         United States Senate   
14119   ANDERSON   Clinton High                         United States Senate   
...          ...            ...                                          ...   
140913    WILSON    Provisional                      United States President   
140916    WILSON    Provisional                         United States Senate   
140917    WILSON    Provisional       U.S. House of Representatives District   
140918    WILSON    Provisional  Tennessee House of Representatives District   
140919    WILSON    Provisional  Tennessee House of Representatives District   

               electdate candidate  vote_count  
14095   November 6, 2012       NaN           0  
14101   November 6, 2012       NaN           0  
14107   November 6, 2012       NaN           0  
14113   November 6, 2012       NaN           0  
14119   November 6, 2012       NaN           0  
...                  ...       ...         ...  
140913  November 6, 2012       NaN           0  
140916  November 6, 2012       NaN           0  
140917  November 6, 2012       NaN           0  
140918  November 6, 2012       NaN           0  
140919  November 6, 2012       NaN           0  

[65572 rows x 6 columns]

In [314]:
# confirm all are 0 counts
blank_candidate_2012['vote_count'].value_counts()

0    65572
Name: vote_count, dtype: int64

In [315]:
# drop all null candidates
tn_2012_namemelt.dropna().reset_index(drop=True)

county       precinct                              officename  \
0      ANDERSON  Andersonville                 United States President   
1      ANDERSON  Andersonville                    United States Senate   
2      ANDERSON  Andersonville                    United States Senate   
3      ANDERSON  Andersonville                    United States Senate   
4      ANDERSON  Andersonville  U.S. House of Representatives District   
...         ...            ...                                     ...   
75343    WILSON           25-2                    United States Senate   
75344    WILSON       Absentee                    United States Senate   
75345    WILSON       Absentee                    United States Senate   
75346    WILSON    Provisional                    United States Senate   
75347    WILSON    Provisional                    United States Senate   

              electdate                      candidate  vote_count  
0      November 6, 2012            . Mitt Romney - (R)        1110  
1      November 6, 2012             . Bob Corker - (R)        1067  
2      November 6, 2012        . Write-In - Larry Crim           0  
3      November 6, 2012    . Write-In - Mary D. Wright           0  
4      November 6, 2012      . Chuck Fleischmann - (R)         996  
...                 ...                            ...         ...  
75343  November 6, 2012  . Write-In - Angelia Stinnett           0  
75344  November 6, 2012  . Write-In - Fred R. Anderson           0  
75345  November 6, 2012  . Write-In - Angelia Stinnett           0  
75346  November 6, 2012  . Write-In - Fred R. Anderson           0  
75347  November 6, 2012  . Write-In - Angelia Stinnett           0  

[75348 rows x 6 columns]

In [316]:
# confirn all null candidates dropped
tn_2012_namemelt[tn_2012_namemelt['candidate'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 65572 entries, 14095 to 140919
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   county      65572 non-null  object
 1   precinct    65572 non-null  object
 2   officename  65572 non-null  object
 3   electdate   65572 non-null  object
 4   candidate   0 non-null      object
 5   vote_count  65572 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 3.5+ MB


In [317]:
tn_2012_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140920 entries, 0 to 140919
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   county      140920 non-null  object
 1   precinct    140920 non-null  object
 2   officename  140920 non-null  object
 3   electdate   140920 non-null  object
 4   candidate   75348 non-null   object
 5   vote_count  140920 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 6.5+ MB


### Look at cadidates with 0 votes, drop

In [318]:
tn_2012_namemelt_0votes=tn_2012_namemelt[tn_2012_namemelt['vote_count'] == 0]
tn_2012_namemelt_0votes

county       precinct                                   officename  \
2       ANDERSON  Andersonville                         United States Senate   
3       ANDERSON  Andersonville                         United States Senate   
8       ANDERSON     Briceville                         United States Senate   
9       ANDERSON     Briceville                         United States Senate   
14      ANDERSON       Bull Run                         United States Senate   
...          ...            ...                                          ...   
140915    WILSON    Provisional                         United States Senate   
140916    WILSON    Provisional                         United States Senate   
140917    WILSON    Provisional       U.S. House of Representatives District   
140918    WILSON    Provisional  Tennessee House of Representatives District   
140919    WILSON    Provisional  Tennessee House of Representatives District   

               electdate                      candidate  vote_count  
2       November 6, 2012        . Write-In - Larry Crim           0  
3       November 6, 2012    . Write-In - Mary D. Wright           0  
8       November 6, 2012        . Write-In - Larry Crim           0  
9       November 6, 2012    . Write-In - Mary D. Wright           0  
14      November 6, 2012        . Write-In - Larry Crim           0  
...                  ...                            ...         ...  
140915  November 6, 2012  . Write-In - Angelia Stinnett           0  
140916  November 6, 2012                            NaN           0  
140917  November 6, 2012                            NaN           0  
140918  November 6, 2012                            NaN           0  
140919  November 6, 2012                            NaN           0  

[97500 rows x 6 columns]

In [319]:
# drop rows with 0 in "vote_count" column
tn_2012_namemelt=tn_2012_namemelt[tn_2012_namemelt['vote_count'] != 0].reset_index(drop=True)
tn_2012_namemelt

county       precinct                                   officename  \
0        ANDERSON  Andersonville                      United States President   
1        ANDERSON  Andersonville                         United States Senate   
2        ANDERSON  Andersonville       U.S. House of Representatives District   
3        ANDERSON  Andersonville  Tennessee House of Representatives District   
4        ANDERSON     Briceville                      United States President   
...           ...            ...                                          ...   
43415  WILLIAMSON           12-1                         United States Senate   
43416  WILLIAMSON           12-2                         United States Senate   
43417  WILLIAMSON           12-3                         United States Senate   
43418      WILSON           16-1                         United States Senate   
43419      WILSON           22-1                         United States Senate   

              electdate                      candidate  vote_count  
0      November 6, 2012            . Mitt Romney - (R)        1110  
1      November 6, 2012             . Bob Corker - (R)        1067  
2      November 6, 2012      . Chuck Fleischmann - (R)         996  
3      November 6, 2012          . Dennis Powers - (R)         998  
4      November 6, 2012            . Mitt Romney - (R)         227  
...                 ...                            ...         ...  
43415  November 6, 2012  . Write-In - Angelia Stinnett           2  
43416  November 6, 2012  . Write-In - Angelia Stinnett          13  
43417  November 6, 2012  . Write-In - Angelia Stinnett           3  
43418  November 6, 2012  . Write-In - Angelia Stinnett           1  
43419  November 6, 2012  . Write-In - Angelia Stinnett           1  

[43420 rows x 6 columns]

In [320]:
tn_2012_namemelt['vote_count'].value_counts()

1       4472
2       3531
3       2819
4       2166
5       1669
        ... 
1766       1
2611       1
1670       1
1638       1
4839       1
Name: vote_count, Length: 2260, dtype: int64

In [321]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2012_int=tn_2012_namemelt.groupby(['officename','county','candidate'], as_index=False)['vote_count'].sum()
tn_2012_int=tn_2012_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2012_int.head(30)

officename      county  \
0                          United States Senate      WILSON   
1                       United States President      WILSON   
2        U.S. House of Representatives District      WILSON   
3   Tennessee House of Representatives District      WILSON   
4                       United States President      WILSON   
5   Tennessee House of Representatives District      WILSON   
6                          United States Senate      WILSON   
7        U.S. House of Representatives District      WILSON   
8        U.S. House of Representatives District      WILSON   
9                          United States Senate      WILSON   
10                         United States Senate      WILSON   
11                         United States Senate      WILSON   
12                      United States President      WILSON   
13                         United States Senate      WILSON   
14                         United States Senate      WILSON   
15                         United States Senate      WILSON   
16                         United States Senate      WILSON   
17                      United States President      WILSON   
18                      United States President      WILSON   
19                      United States President      WILSON   
20                      United States President      WILSON   
21       U.S. House of Representatives District      WILSON   
22                         United States Senate      WILSON   
23                         United States Senate      WILSON   
24                         United States Senate      WILSON   
25                         United States Senate      WILSON   
26                         United States Senate  WILLIAMSON   
27       U.S. House of Representatives District  WILLIAMSON   
28                      United States President  WILLIAMSON   
29                      United States President  WILLIAMSON   

                             candidate  vote_count  
0                   . Bob Corker - (R)       36993  
1                  . Mitt Romney - (R)       36109  
2                  . Diane Black - (R)       35659  
3                   . Susan Lynn - (R)       24116  
4                 . Barack Obama - (D)       14695  
5                 . Mark A. Pody - (R)       12731  
6              . Mark E. Clayton - (D)       10497  
7                . Scott Beasley - (I)        7101  
8                    . Pat Riley - (G)        4474  
9              . Martin Pleasant - (G)         749  
10                . Kermit Steck - (C)         512  
11            . Shaun E. Crowell - (I)         450  
12                . Gary Johnson - (I)         381  
13        . Troy Stephen Scoggin - (I)         176  
14                . James Higdon - (I)         136  
15          . Michel Joseph Long - (I)         131  
16              . David Gatchell - (I)         121  
17                . Virgil Goode - (C)         117  
18                  . Jill Stein - (G)         113  
19    . Ross C. "Rocky" Anderson - (I)          43  
20               . Merlin Miller - (I)          41  
21        . Write-In - Rachel Robinson          13  
22           . Write-In - Jacob Maurer           7  
23             . Write-In - Larry Crim           7  
24  . Write-In - Christopher Schappert           3  
25       . Write-In - Angelia Stinnett           2  
26                  . Bob Corker - (R)       72402  
27            . Marsha Blackburn - (R)       71938  
28                 . Mitt Romney - (R)       69850  
29                . Barack Obama - (D)       25142

In [322]:
# Capitalize couty names
tn_2012_int['county']=tn_2012_int['county'].str.capitalize()
tn_2012_int.head(5)

officename  county             candidate  \
0                         United States Senate  Wilson    . Bob Corker - (R)   
1                      United States President  Wilson   . Mitt Romney - (R)   
2       U.S. House of Representatives District  Wilson   . Diane Black - (R)   
3  Tennessee House of Representatives District  Wilson    . Susan Lynn - (R)   
4                      United States President  Wilson  . Barack Obama - (D)   

   vote_count  
0       36993  
1       36109  
2       35659  
3       24116  
4       14695

In [323]:
tn_2012_int.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2445 entries, 0 to 2444
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   officename  2445 non-null   object
 1   county      2445 non-null   object
 2   candidate   2445 non-null   object
 3   vote_count  2445 non-null   int64 
dtypes: int64(1), object(3)
memory usage: 76.5+ KB


In [324]:
tn_2012_int['officename'].unique()

array(['United States Senate', 'United States President',
       'U.S. House of Representatives District',
       'Tennessee House of Representatives District',
       'Tennessee Senate District'], dtype=object)

### Clean up names and greate separate party column

In [325]:
# new dataframe new_names of non-write-in candidates
new_names=tn_2012_int[tn_2012_int['candidate'].str.contains("Write")==False]
new_names

officename    county  \
0                            United States Senate    Wilson   
1                         United States President    Wilson   
2          U.S. House of Representatives District    Wilson   
3     Tennessee House of Representatives District    Wilson   
4                         United States President    Wilson   
...                                           ...       ...   
2435                      United States President  Anderson   
2436                         United States Senate  Anderson   
2437                         United States Senate  Anderson   
2438                      United States President  Anderson   
2439                      United States President  Anderson   

                             candidate  vote_count  
0                   . Bob Corker - (R)       36993  
1                  . Mitt Romney - (R)       36109  
2                  . Diane Black - (R)       35659  
3                   . Susan Lynn - (R)       24116  
4                 . Barack Obama - (D)       14695  
...                                ...         ...  
2435                . Jill Stein - (G)          88  
2436      . Troy Stephen Scoggin - (I)          72  
2437            . David Gatchell - (I)          60  
2438  . Ross C. "Rocky" Anderson - (I)          54  
2439             . Merlin Miller - (I)          25  

[2311 rows x 4 columns]

In [326]:
#drop period in front of names
new_names['candidate']=new_names['candidate'].str[1:]
new_names

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


officename    county  \
0                            United States Senate    Wilson   
1                         United States President    Wilson   
2          U.S. House of Representatives District    Wilson   
3     Tennessee House of Representatives District    Wilson   
4                         United States President    Wilson   
...                                           ...       ...   
2435                      United States President  Anderson   
2436                         United States Senate  Anderson   
2437                         United States Senate  Anderson   
2438                      United States President  Anderson   
2439                      United States President  Anderson   

                            candidate  vote_count  
0                    Bob Corker - (R)       36993  
1                   Mitt Romney - (R)       36109  
2                   Diane Black - (R)       35659  
3                    Susan Lynn - (R)       24116  
4                  Barack Obama - (D)       14695  
...                               ...         ...  
2435                 Jill Stein - (G)          88  
2436       Troy Stephen Scoggin - (I)          72  
2437             David Gatchell - (I)          60  
2438   Ross C. "Rocky" Anderson - (I)          54  
2439              Merlin Miller - (I)          25  

[2311 rows x 4 columns]

In [327]:
# split name fromm party desination
new_names=tn_2012_int_name['candidate'].str.split(pat='-',expand=True)
new_names

0     1
0                    Bob Corker    (R)
1                   Mitt Romney    (R)
2                   Diane Black    (R)
3                    Susan Lynn    (R)
4                  Barack Obama    (D)
...                          ...   ...
2435                 Jill Stein    (G)
2436       Troy Stephen Scoggin    (I)
2437             David Gatchell    (I)
2438   Ross C. "Rocky" Anderson    (I)
2439              Merlin Miller    (I)

[2311 rows x 2 columns]

In [391]:
new_names.rename(columns={0:'name',1:'party'},inplace=True)
new_names

name party
0                    Bob Corker    (R)
1                   Mitt Romney    (R)
2                   Diane Black    (R)
3                    Susan Lynn    (R)
4                  Barack Obama    (D)
...                          ...   ...
2435                 Jill Stein    (G)
2436       Troy Stephen Scoggin    (I)
2437             David Gatchell    (I)
2438   Ross C. "Rocky" Anderson    (I)
2439              Merlin Miller    (I)

[2311 rows x 2 columns]

In [392]:
# new dataframe tn_2012_int_wi of only write-in candidates
tn_2012_int_wi=tn_2012_int[tn_2012_int['candidate'].str.contains("\(")==False]
tn_2012_int_wi

officename    county  \
21    U.S. House of Representatives District    Wilson   
22                      United States Senate    Wilson   
23                      United States Senate    Wilson   
24                      United States Senate    Wilson   
25                      United States Senate    Wilson   
...                                      ...       ...   
2440  U.S. House of Representatives District  Anderson   
2441                    United States Senate  Anderson   
2442                    United States Senate  Anderson   
2443                    United States Senate  Anderson   
2444                    United States Senate  Anderson   

                               candidate  vote_count  
21          . Write-In - Rachel Robinson          13  
22             . Write-In - Jacob Maurer           7  
23               . Write-In - Larry Crim           7  
24    . Write-In - Christopher Schappert           3  
25         . Write-In - Angelia Stinnett           2  
...                                  ...         ...  
2440           . Write-In - Richard Hall           5  
2441           . Write-In - Jacob Maurer           5  
2442            . Write-In - Jim Maynard           2  
2443       . Write-In - Angelia Stinnett           1  
2444             . Write-In - Larry Crim           1  

[134 rows x 4 columns]

In [393]:
# split name fromm 'Write-in'
new_wi_name=tn_2012_int_wi['candidate'].str.split(pat='-',n=2,expand=True)
new_wi_name

0    1                       2
21    . Write  In          Rachel Robinson
22    . Write  In             Jacob Maurer
23    . Write  In               Larry Crim
24    . Write  In    Christopher Schappert
25    . Write  In         Angelia Stinnett
...       ...  ...                     ...
2440  . Write  In             Richard Hall
2441  . Write  In             Jacob Maurer
2442  . Write  In              Jim Maynard
2443  . Write  In         Angelia Stinnett
2444  . Write  In               Larry Crim

[134 rows x 3 columns]

In [394]:
# create new 'party' columnn with "Write-In"
new_wi_name['party']=new_wi_name[[0,1]].agg('-'.join,axis=1).str[2:]
new_wi_name

0    1                       2      party
21    . Write  In          Rachel Robinson  Write-In 
22    . Write  In             Jacob Maurer  Write-In 
23    . Write  In               Larry Crim  Write-In 
24    . Write  In    Christopher Schappert  Write-In 
25    . Write  In         Angelia Stinnett  Write-In 
...       ...  ...                     ...        ...
2440  . Write  In             Richard Hall  Write-In 
2441  . Write  In             Jacob Maurer  Write-In 
2442  . Write  In              Jim Maynard  Write-In 
2443  . Write  In         Angelia Stinnett  Write-In 
2444  . Write  In               Larry Crim  Write-In 

[134 rows x 4 columns]

In [395]:
#drop columns 0 and 1, rename column 2 to"candidate"
new_wi_name=new_wi_name.drop([0,1],axis=1)
new_wi_name

2      party
21           Rachel Robinson  Write-In 
22              Jacob Maurer  Write-In 
23                Larry Crim  Write-In 
24     Christopher Schappert  Write-In 
25          Angelia Stinnett  Write-In 
...                      ...        ...
2440            Richard Hall  Write-In 
2441            Jacob Maurer  Write-In 
2442             Jim Maynard  Write-In 
2443        Angelia Stinnett  Write-In 
2444              Larry Crim  Write-In 

[134 rows x 2 columns]

In [413]:
new_wi_name.rename(columns={2:'name'},inplace=True)
new_wi_name

name      party
21           Rachel Robinson  Write-In 
22              Jacob Maurer  Write-In 
23                Larry Crim  Write-In 
24     Christopher Schappert  Write-In 
25          Angelia Stinnett  Write-In 
...                      ...        ...
2440            Richard Hall  Write-In 
2441            Jacob Maurer  Write-In 
2442             Jim Maynard  Write-In 
2443        Angelia Stinnett  Write-In 
2444              Larry Crim  Write-In 

[134 rows x 2 columns]

In [425]:
# join new_names and new_wi_names
frames=[new_names,new_wi_name]
cleaned_names=pd.concat(frames)
cleaned_names

name      party
0           Bob Corker         (R)
1          Mitt Romney         (R)
2          Diane Black         (R)
3           Susan Lynn         (R)
4         Barack Obama         (D)
...                 ...        ...
2440       Richard Hall  Write-In 
2441       Jacob Maurer  Write-In 
2442        Jim Maynard  Write-In 
2443   Angelia Stinnett  Write-In 
2444         Larry Crim  Write-In 

[2445 rows x 2 columns]

### Join cleaned names and party column to tn_2012_int

In [426]:
name=cleaned_names['name']
party=cleaned_names['party']
tn_2012_final=tn_2012_int.join(name).join(party)
tn_2012_final

officename    county  \
0                            United States Senate    Wilson   
1                         United States President    Wilson   
2          U.S. House of Representatives District    Wilson   
3     Tennessee House of Representatives District    Wilson   
4                         United States President    Wilson   
...                                           ...       ...   
2440       U.S. House of Representatives District  Anderson   
2441                         United States Senate  Anderson   
2442                         United States Senate  Anderson   
2443                         United States Senate  Anderson   
2444                         United States Senate  Anderson   

                          candidate  vote_count               name      party  
0                . Bob Corker - (R)       36993        Bob Corker         (R)  
1               . Mitt Romney - (R)       36109       Mitt Romney         (R)  
2               . Diane Black - (R)       35659       Diane Black         (R)  
3                . Susan Lynn - (R)       24116        Susan Lynn         (R)  
4              . Barack Obama - (D)       14695      Barack Obama         (D)  
...                             ...         ...                ...        ...  
2440      . Write-In - Richard Hall           5       Richard Hall  Write-In   
2441      . Write-In - Jacob Maurer           5       Jacob Maurer  Write-In   
2442       . Write-In - Jim Maynard           2        Jim Maynard  Write-In   
2443  . Write-In - Angelia Stinnett           1   Angelia Stinnett  Write-In   
2444        . Write-In - Larry Crim           1         Larry Crim  Write-In   

[2445 rows x 6 columns]

In [427]:
tn_2012_final['party'].value_counts()

 (I)         940
 (R)         504
 (D)         421
 (G)         256
 (C)         190
Write-In     134
Name: party, dtype: int64

In [428]:
# change "(R)" to "Republican" and "(D)"to "Democratic" in "party" column
tn_2012_final["party"]=tn_2012_final['party'].replace('(R)',"Republican")
tn_2012_final["party"]=tn_2012_final['party'].str.replace('\(R\)',"Republican").str.replace('\(D\)',"Democratic")
tn_2012_final

officename    county  \
0                            United States Senate    Wilson   
1                         United States President    Wilson   
2          U.S. House of Representatives District    Wilson   
3     Tennessee House of Representatives District    Wilson   
4                         United States President    Wilson   
...                                           ...       ...   
2440       U.S. House of Representatives District  Anderson   
2441                         United States Senate  Anderson   
2442                         United States Senate  Anderson   
2443                         United States Senate  Anderson   
2444                         United States Senate  Anderson   

                          candidate  vote_count               name  \
0                . Bob Corker - (R)       36993        Bob Corker    
1               . Mitt Romney - (R)       36109       Mitt Romney    
2               . Diane Black - (R)       35659       Diane Black    
3                . Susan Lynn - (R)       24116        Susan Lynn    
4              . Barack Obama - (D)       14695      Barack Obama    
...                             ...         ...                ...   
2440      . Write-In - Richard Hall           5       Richard Hall   
2441      . Write-In - Jacob Maurer           5       Jacob Maurer   
2442       . Write-In - Jim Maynard           2        Jim Maynard   
2443  . Write-In - Angelia Stinnett           1   Angelia Stinnett   
2444        . Write-In - Larry Crim           1         Larry Crim   

            party  
0      Republican  
1      Republican  
2      Republican  
3      Republican  
4      Democratic  
...           ...  
2440    Write-In   
2441    Write-In   
2442    Write-In   
2443    Write-In   
2444    Write-In   

[2445 rows x 6 columns]

In [432]:
# create data frame tn_2012_pres for presidental election
tn_2012_pres=tn_2012_final[tn_2012_final['officename'].str.match('United States President')]
tn_2012_pres=tn_2012_pres.reset_index(drop=True)
tn_2012_pres

officename    county                         candidate  \
0    United States President    Wilson               . Mitt Romney - (R)   
1    United States President    Wilson              . Barack Obama - (D)   
2    United States President    Wilson              . Gary Johnson - (I)   
3    United States President    Wilson              . Virgil Goode - (C)   
4    United States President    Wilson                . Jill Stein - (G)   
..                       ...       ...                               ...   
659  United States President  Anderson              . Gary Johnson - (I)   
660  United States President  Anderson              . Virgil Goode - (C)   
661  United States President  Anderson                . Jill Stein - (G)   
662  United States President  Anderson  . Ross C. "Rocky" Anderson - (I)   
663  United States President  Anderson             . Merlin Miller - (I)   

     vote_count                        name        party  
0         36109                Mitt Romney    Republican  
1         14695               Barack Obama    Democratic  
2           381               Gary Johnson           (I)  
3           117               Virgil Goode           (C)  
4           113                 Jill Stein           (G)  
..          ...                         ...          ...  
659         301               Gary Johnson           (I)  
660         101               Virgil Goode           (C)  
661          88                 Jill Stein           (G)  
662          54   Ross C. "Rocky" Anderson           (I)  
663          25              Merlin Miller           (I)  

[664 rows x 6 columns]

In [433]:
tn_2012_pres.head(20)

officename      county                         candidate  \
0   United States President      Wilson               . Mitt Romney - (R)   
1   United States President      Wilson              . Barack Obama - (D)   
2   United States President      Wilson              . Gary Johnson - (I)   
3   United States President      Wilson              . Virgil Goode - (C)   
4   United States President      Wilson                . Jill Stein - (G)   
5   United States President      Wilson  . Ross C. "Rocky" Anderson - (I)   
6   United States President      Wilson             . Merlin Miller - (I)   
7   United States President  Williamson               . Mitt Romney - (R)   
8   United States President  Williamson              . Barack Obama - (D)   
9   United States President  Williamson              . Gary Johnson - (I)   
10  United States President  Williamson                . Jill Stein - (G)   
11  United States President  Williamson              . Virgil Goode - (C)   
12  United States President  Williamson  . Ross C. "Rocky" Anderson - (I)   
13  United States President  Williamson             . Merlin Miller - (I)   
14  United States President       White               . Mitt Romney - (R)   
15  United States President       White              . Barack Obama - (D)   
16  United States President       White              . Gary Johnson - (I)   
17  United States President       White              . Virgil Goode - (C)   
18  United States President       White                . Jill Stein - (G)   
19  United States President       White  . Ross C. "Rocky" Anderson - (I)   

    vote_count                        name        party  
0        36109                Mitt Romney    Republican  
1        14695               Barack Obama    Democratic  
2          381               Gary Johnson           (I)  
3          117               Virgil Goode           (C)  
4          113                 Jill Stein           (G)  
5           43   Ross C. "Rocky" Anderson           (I)  
6           41              Merlin Miller           (I)  
7        69850                Mitt Romney    Republican  
8        25142               Barack Obama    Democratic  
9          748               Gary Johnson           (I)  
10         217                 Jill Stein           (G)  
11         160               Virgil Goode           (C)  
12          66   Ross C. "Rocky" Anderson           (I)  
13          42              Merlin Miller           (I)  
14        6197                Mitt Romney    Republican  
15        2795               Barack Obama    Democratic  
16          68               Gary Johnson           (I)  
17          37               Virgil Goode           (C)  
18          22                 Jill Stein           (G)  
19          11   Ross C. "Rocky" Anderson           (I)

### Add column that has total vote count by count for use in calculating county vote percentages

In [434]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2012_pres.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0      Wilson       36109
1      Wilson       14695
2      Wilson         381
3      Wilson         117
4      Wilson         113
..        ...         ...
659  Anderson         301
660  Anderson         101
661  Anderson          88
662  Anderson          54
663  Anderson          25

[664 rows x 2 columns]

In [435]:
#sum county votes
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson      29659
Bedford       14445
Benton         6226
Bledsoe        4359
Blount        49234
              ...  
Wayne          5486
Weakley       12337
White          9140
Williamson    96225
Wilson        51499
Name: vote_count, Length: 95, dtype: int64

In [436]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2016_pres correlating with value in each row
tn_2012_pres['sum_county_vote']=tn_2012_pres['county'].map(county_vote_sum)
tn_2012_pres.head(20)

officename      county                         candidate  \
0   United States President      Wilson               . Mitt Romney - (R)   
1   United States President      Wilson              . Barack Obama - (D)   
2   United States President      Wilson              . Gary Johnson - (I)   
3   United States President      Wilson              . Virgil Goode - (C)   
4   United States President      Wilson                . Jill Stein - (G)   
5   United States President      Wilson  . Ross C. "Rocky" Anderson - (I)   
6   United States President      Wilson             . Merlin Miller - (I)   
7   United States President  Williamson               . Mitt Romney - (R)   
8   United States President  Williamson              . Barack Obama - (D)   
9   United States President  Williamson              . Gary Johnson - (I)   
10  United States President  Williamson                . Jill Stein - (G)   
11  United States President  Williamson              . Virgil Goode - (C)   
12  United States President  Williamson  . Ross C. "Rocky" Anderson - (I)   
13  United States President  Williamson             . Merlin Miller - (I)   
14  United States President       White               . Mitt Romney - (R)   
15  United States President       White              . Barack Obama - (D)   
16  United States President       White              . Gary Johnson - (I)   
17  United States President       White              . Virgil Goode - (C)   
18  United States President       White                . Jill Stein - (G)   
19  United States President       White  . Ross C. "Rocky" Anderson - (I)   

    vote_count                        name        party  sum_county_vote  
0        36109                Mitt Romney    Republican            51499  
1        14695               Barack Obama    Democratic            51499  
2          381               Gary Johnson           (I)            51499  
3          117               Virgil Goode           (C)            51499  
4          113                 Jill Stein           (G)            51499  
5           43   Ross C. "Rocky" Anderson           (I)            51499  
6           41              Merlin Miller           (I)            51499  
7        69850                Mitt Romney    Republican            96225  
8        25142               Barack Obama    Democratic            96225  
9          748               Gary Johnson           (I)            96225  
10         217                 Jill Stein           (G)            96225  
11         160               Virgil Goode           (C)            96225  
12          66   Ross C. "Rocky" Anderson           (I)            96225  
13          42              Merlin Miller           (I)            96225  
14        6197                Mitt Romney    Republican             9140  
15        2795               Barack Obama    Democratic             9140  
16          68               Gary Johnson           (I)             9140  
17          37               Virgil Goode           (C)             9140  
18          22                 Jill Stein           (G)             9140  
19          11   Ross C. "Rocky" Anderson           (I)             9140

In [438]:
# create "pct_county_vote" column for tn_2016_pres
tn_2012_pres['pct_county_vote']=round(tn_2012_pres['vote_count']/tn_2012_pres['sum_county_vote']*100,2)
tn_2012_pres.head(20)

officename      county                         candidate  \
0   United States President      Wilson               . Mitt Romney - (R)   
1   United States President      Wilson              . Barack Obama - (D)   
2   United States President      Wilson              . Gary Johnson - (I)   
3   United States President      Wilson              . Virgil Goode - (C)   
4   United States President      Wilson                . Jill Stein - (G)   
5   United States President      Wilson  . Ross C. "Rocky" Anderson - (I)   
6   United States President      Wilson             . Merlin Miller - (I)   
7   United States President  Williamson               . Mitt Romney - (R)   
8   United States President  Williamson              . Barack Obama - (D)   
9   United States President  Williamson              . Gary Johnson - (I)   
10  United States President  Williamson                . Jill Stein - (G)   
11  United States President  Williamson              . Virgil Goode - (C)   
12  United States President  Williamson  . Ross C. "Rocky" Anderson - (I)   
13  United States President  Williamson             . Merlin Miller - (I)   
14  United States President       White               . Mitt Romney - (R)   
15  United States President       White              . Barack Obama - (D)   
16  United States President       White              . Gary Johnson - (I)   
17  United States President       White              . Virgil Goode - (C)   
18  United States President       White                . Jill Stein - (G)   
19  United States President       White  . Ross C. "Rocky" Anderson - (I)   

    vote_count                        name        party  sum_county_vote  \
0        36109                Mitt Romney    Republican            51499   
1        14695               Barack Obama    Democratic            51499   
2          381               Gary Johnson           (I)            51499   
3          117               Virgil Goode           (C)            51499   
4          113                 Jill Stein           (G)            51499   
5           43   Ross C. "Rocky" Anderson           (I)            51499   
6           41              Merlin Miller           (I)            51499   
7        69850                Mitt Romney    Republican            96225   
8        25142               Barack Obama    Democratic            96225   
9          748               Gary Johnson           (I)            96225   
10         217                 Jill Stein           (G)            96225   
11         160               Virgil Goode           (C)            96225   
12          66   Ross C. "Rocky" Anderson           (I)            96225   
13          42              Merlin Miller           (I)            96225   
14        6197                Mitt Romney    Republican             9140   
15        2795               Barack Obama    Democratic             9140   
16          68               Gary Johnson           (I)             9140   
17          37               Virgil Goode           (C)             9140   
18          22                 Jill Stein           (G)             9140   
19          11   Ross C. "Rocky" Anderson           (I)             9140   

    pct_county_vote  
0             70.12  
1             28.53  
2              0.74  
3              0.23  
4              0.22  
5              0.08  
6              0.08  
7             72.59  
8             26.13  
9              0.78  
10             0.23  
11             0.17  
12             0.07  
13             0.04  
14            67.80  
15            30.58  
16             0.74  
17             0.40  
18             0.24  
19             0.12

In [440]:
# how many parties?
tn_2012_pres['party'].value_counts()

 (I)           284
 Democratic     95
 (C)            95
 Republican     95
 (G)            95
Name: party, dtype: int64

## Create dataframe "tn_2012_county_pres_pct"  with percent values per party per county

In [443]:
# create new dataframe with reduced columns values only
tn_2012_pct_county=tn_2012_pres.loc[0:,['county','party','pct_county_vote']]
tn_2012_pct_county

county        party  pct_county_vote
0      Wilson   Republican            70.12
1      Wilson   Democratic            28.53
2      Wilson          (I)             0.74
3      Wilson          (C)             0.23
4      Wilson          (G)             0.22
..        ...          ...              ...
659  Anderson          (I)             1.01
660  Anderson          (C)             0.34
661  Anderson          (G)             0.30
662  Anderson          (I)             0.18
663  Anderson          (I)             0.08

[664 rows x 3 columns]

In [449]:
# create new dataframe with Republican values only
tn_2012_rep_pct=tn_2012_pct_county[tn_2012_pct_county['party']== " Republican"].reset_index(drop=True)
tn_2012_rep_pct

county        party  pct_county_vote
0       Wilson   Republican            70.12
1   Williamson   Republican            72.59
2        White   Republican            67.80
3      Weakley   Republican            69.75
4        Wayne   Republican            77.52
..         ...          ...              ...
90      Blount   Republican            71.98
91     Bledsoe   Republican            69.33
92      Benton   Republican            61.84
93     Bedford   Republican            69.46
94    Anderson   Republican            63.95

[95 rows x 3 columns]

In [450]:
# rename 'pct_county_vote' to 'rep_pct' and drop "party" column to prepare for merge 
tn_2012_rep_pct=tn_2012_rep_pct.rename(columns={'pct_county_vote':'Republican'},inplace=False).drop(['party'], axis=1)
tn_2012_rep_pct

county  Republican
0       Wilson       70.12
1   Williamson       72.59
2        White       67.80
3      Weakley       69.75
4        Wayne       77.52
..         ...         ...
90      Blount       71.98
91     Bledsoe       69.33
92      Benton       61.84
93     Bedford       69.46
94    Anderson       63.95

[95 rows x 2 columns]

In [452]:
# create new dataframe with Democratic values only
tn_2012_dem_pct=tn_2012_pct_county[tn_2012_pct_county['party']== " Democratic"].reset_index(drop=True)
tn_2012_dem_pct

county        party  pct_county_vote
0       Wilson   Democratic            28.53
1   Williamson   Democratic            26.13
2        White   Democratic            30.58
3      Weakley   Democratic            28.76
4        Wayne   Democratic            21.20
..         ...          ...              ...
90      Blount   Democratic            26.27
91     Bledsoe   Democratic            29.07
92      Benton   Democratic            36.27
93     Bedford   Democratic            29.15
94    Anderson   Democratic            34.13

[95 rows x 3 columns]

In [453]:
# rename 'pct_county_vote' to 'dem_pct' and drop "party" to prepare for merge 
tn_2012_dem_pct=tn_2012_dem_pct.rename(columns={'pct_county_vote':'Democratic'},inplace=False).drop(['party'], axis=1)
tn_2012_dem_pct

county  Democratic
0       Wilson       28.53
1   Williamson       26.13
2        White       30.58
3      Weakley       28.76
4        Wayne       21.20
..         ...         ...
90      Blount       26.27
91     Bledsoe       29.07
92      Benton       36.27
93     Bedford       29.15
94    Anderson       34.13

[95 rows x 2 columns]

In [456]:
# create new dataframe with non Republican or Democratic values only
tn_2012_other_pct=tn_2012_pct_county[(tn_2012_pct_county['party']!=" Republican")&(tn_2012_pct_county['party']!=" Democratic")].reset_index(drop=True)
tn_2012_other_pct

county party  pct_county_vote
0      Wilson   (I)             0.74
1      Wilson   (C)             0.23
2      Wilson   (G)             0.22
3      Wilson   (I)             0.08
4      Wilson   (I)             0.08
..        ...   ...              ...
469  Anderson   (I)             1.01
470  Anderson   (C)             0.34
471  Anderson   (G)             0.30
472  Anderson   (I)             0.18
473  Anderson   (I)             0.08

[474 rows x 3 columns]

In [457]:
# sum by county
tn_2012_other_pct=tn_2012_other_pct.groupby(['county'],as_index=False)['pct_county_vote'].sum()
tn_2012_other_pct

county  pct_county_vote
0     Anderson             1.91
1      Bedford             1.38
2       Benton             1.89
3      Bledsoe             1.60
4       Blount             1.74
..         ...              ...
90       Wayne             1.28
91     Weakley             1.49
92       White             1.61
93  Williamson             1.29
94      Wilson             1.35

[95 rows x 2 columns]

In [458]:
# rename 'pct_county_vote' to 'other_pct' and drop "party" to prepare for merge 
tn_2012_other_pct=tn_2012_other_pct.rename(columns={'pct_county_vote':'Other'})
tn_2012_other_pct

county  Other
0     Anderson   1.91
1      Bedford   1.38
2       Benton   1.89
3      Bledsoe   1.60
4       Blount   1.74
..         ...    ...
90       Wayne   1.28
91     Weakley   1.49
92       White   1.61
93  Williamson   1.29
94      Wilson   1.35

[95 rows x 2 columns]

In [459]:
# join tn_2012_rep, tn_2012_dem_pct and tn_2012_other_pct to form tn_2012_county_pres_pct
dem=tn_2012_dem_pct['Democratic']
other=tn_2012_other_pct['Other']
tn_2012_county_pres_pct=tn_2012_rep_pct.join(dem).join(other)
tn_2012_county_pres_pct

county  Republican  Democratic  Other
0       Wilson       70.12       28.53   1.91
1   Williamson       72.59       26.13   1.38
2        White       67.80       30.58   1.89
3      Weakley       69.75       28.76   1.60
4        Wayne       77.52       21.20   1.74
..         ...         ...         ...    ...
90      Blount       71.98       26.27   1.28
91     Bledsoe       69.33       29.07   1.49
92      Benton       61.84       36.27   1.61
93     Bedford       69.46       29.15   1.29
94    Anderson       63.95       34.13   1.35

[95 rows x 4 columns]

### Add column for Republican margin over Democrat for heat mapping in Tableau

In [460]:
tn_2012_county_pres_pct['Rep_Margin']= tn_2012_county_pres_pct['Republican']-tn_2012_county_pres_pct['Democratic']
tn_2012_county_pres_pct.head(25)

county  Republican  Democratic  Other  Rep_Margin
0       Wilson       70.12       28.53   1.91       41.59
1   Williamson       72.59       26.13   1.38       46.46
2        White       67.80       30.58   1.89       37.22
3      Weakley       69.75       28.76   1.60       40.99
4        Wayne       77.52       21.20   1.74       56.32
5   Washington       68.30       29.82   1.40       38.48
6       Warren       61.54       36.51   1.40       25.03
7    Van buren       60.26       38.04   2.00       22.22
8        Union       73.47       25.36   1.39       48.11
9       Unicoi       71.01       27.00   1.58       44.01
10   Trousdale       55.49       42.69   1.68       12.80
11      Tipton       69.23       29.62   1.58       39.61
12      Sumner       70.38       28.42   1.27       41.96
13    Sullivan       72.74       25.58   1.28       47.16
14     Stewart       57.93       40.45   1.66       17.48
15       Smith       63.43       34.85   1.91       28.58
16      Shelby       36.55       62.63   0.83      -26.08
17      Sevier       76.73       21.91   1.32       54.82
18  Sequatchie       68.86       28.96   1.69       39.90
19       Scott       76.74       21.78   1.75       54.96
20  Rutherford       61.56       36.84   1.64       24.72
21   Robertson       67.11       31.53   1.71       35.58
22       Roane       69.82       28.53   0.89       41.29
23        Rhea       73.67       24.82   1.01       48.85
24      Putnam       67.66       30.60   1.33       37.06

In [461]:
tn_2012_county_pres_pct.to_excel(r'../data/2012_election/tn_2012_county_pres_pct.xlsx',index=False)